packages/commonwealth/client/scripts/views/pages/general_contract/index.tsx

```diff
@@ -0,0 +1,226 @@
+/* @jsx m */
+
+import 'pages/general_contract/index.scss';
+import app from 'state';
+import { Contract } from 'models';
+import m from 'mithril';
+import ClassComponent from 'class_component';
+
+import { notifyError, notifySuccess } from 'controllers/app/notifications';
+import { AbiItem, AbiInput, AbiOutput } from 'web3-utils/types';
+import { CWText } from 'views/components/component_kit/cw_text';
+import { CWButton } from 'views/components/component_kit/cw_button';
+import { CWTextInput } from 'views/components/component_kit/cw_text_input';
+import { ChainBase } from 'common-common/src/types';
+import { TransactionReceipt } from 'web3-core';
+import { parseFunctionsFromABI } from 'helpers/abi_utils';
+import GeneralContractsController from 'controllers/chain/ethereum/generalContracts';
+import {
+  handleMappingAbiInputs,
+  processAbiInputsToDataTypes,
+  validateAbiInput,
+} from 'helpers/abi_form_helpers';
+import { CWSpinner } from 'views/components/component_kit/cw_spinner';
+import { PageNotFound } from '../404';
+import { PageLoading } from '../loading';
+import Sublayout from '../../sublayout';
+import { ChainFormState } from '../create_community/types';
+
+type CreateContractForm = {
+  functionNameToFunctionInputArgs: Map<string, Map<number, string>>;
+};
+
+type CreateContractState = ChainFormState & {
+  functionNameToFunctionOutput: Map<string, any[]>;
+  form: CreateContractForm;
+};
+class GeneralContractPage
+  extends ClassComponent<{ contractAddress?: string }>
+{
+  generalContractsController: GeneralContractsController;
+  private state: CreateContractState = {
+    message: '',
+    loaded: false,
+    loading: false,
+    saving: false,
+    status: undefined,
+    functionNameToFunctionOutput: new Map<string, any[]>(),
+    form: {
+      functionNameToFunctionInputArgs: new Map<string, Map<number, string>>(),
+    },
+  };
+
+  view(vnode) {
+    const fetchContractAbi = async (contract: Contract) => {
+      if (contract.abi === undefined) {
+        // use the contract address to fetch the abi using controller
+        await app.contracts.checkFetchEtherscanForAbi(contract.address);
+        // TODO The UI Should In One Go show the abi form after successfully fetching the abi
+        // from etherscan, which it does not do rn
+        m.redraw();
+      }
+    };
+
+    const callFunction = async (contractAddress: string, fn: AbiItem) => {
+      try {
+        this.state.loading = true;
+        // initialize daoFactory Controller
+        this.generalContractsController = new GeneralContractsController(
```

Comment:
Should this be a singleton? Or a class with static methods? This init flow doesn't make a lot of sense.